### PR TITLE
Render display names of users and rooms according to the spec - take 2

### PR DESCRIPTION
**Pull request by KitsuneRal**
Though it's yet incomplete (room names are on the way), I think you can start reviewing this. As you asked, I have split one big commit into smaller ones (and this is only the lib part, naturally). The code is backwards-compatible; however, I had to lay some leftovers for that - namely, User::displayname() and User::nameChanged(). Probably, they should be marked as deprecated in the /** docs */. I'll submit a separate PR on Quaternion later to reflect the newly introduced functionality in the UI.

### CODE REVIEW

**Comment by KitsuneRal** [comment]
Room names have arrived. This is a complete PR now.
Upon some more consideration - probably, User::displayname() should remain for cases outside of specific rooms. User::nameChanged is definitely a potential source of performance problems, so I'm deprecating it with the last commit.

**Comment by KitsuneRal** [comment]
Revisiting the topic of user renaming and moving the signal to Connection:
1. User::avatarChanged() is in the same position as User::nameChanged(), so the approach should be common for both.
2. When moving the signal to Connection, as I've just realized, UserListModel will be spammed with unneeded signals from non-members of the room (in case such events arrive, though - but I wouldn't rule this out). On the other hand, this provides additional flexibility - we might want to use UserListModel, e.g., to display all users on the server at some point in time; the room context is irrelevant in such case.
3. Using a signal on User level makes perfect sense when showing a window for that user. Sorry that I run too fast; User::nameChanged/avatarChanged should stay as well. I guess the right way would be to connect User::nameChanged to Connection::userRenamed in that case.

Therefore I'll remove the last commit (about deprecating User::nameChanged), connect the two signals and add Connection::avatarChanged "mediation signal" in a similar fashion - in the same PR, if you don't mind.

**Comment by KitsuneRal** [comment]
Just a notice: because I force-pushed the branch, some commits shifted later than comments referring to them. In case you already started reviewing - I amended [8d56fcf](
https://github.com/Fxrh/libqmatrixclient/pull/4/commits/8d56fcf0046bb192fc8bd6b0fb8166cc62179eef) to drop emitting two signals.

File: connectionprivate.cpp [hunk 1]
```diff
@@ -48,43 +48,46 @@ ConnectionPrivate::~ConnectionPrivate()
...
+
+Room* ConnectionPrivate::provideRoom(QString id)
+{
+    if (id.isEmpty())
+    {
+        qDebug() << "ConnectionPrivate::provideRoom() with empty id, doing nothing";
+        return nullptr;
     }
+
+    Room* room = roomMap.value(id, nullptr);
```
Comment (Fxrh) [inline connectionprivate.cpp]: I think it would be easier to read this way:
```
if( roomMap.contains(id) )
    return roomMap.value(id);
```
It might be a bit slower than your version, but I don't think it matters much (it's a hashmap, so we have O(1) complexity)

What do you think?

File: room.cpp [hunk 2]
```diff
@@ -460,6 +453,95 @@ void Room::processEphemeralEvent(Event* event)
...
+    return QString();
+}
+
+void Room::Private::updateDisplayname()
+{
+    const QString old_name = displayname;
+
+    // CS spec, section 11.2.2.5 Calculating the display name for a room
+    // Numbers and i's below refer to respective parts in the spec.
+    do {
```
Comment (Fxrh) [inline room.cpp]: This do {} while(false) feels a bit like a hack. Can't we do this with if/else if?

**Comment by Fxrh** [comment]
FIrst: Thanks for the PR and sorry it took so long!

I'm not quite sure about moving the nameChanged and avatarChanged signals to the Connection class. Imho that belongs to the User class and I'd like to make sure Connection doesn't evolve into a god class (it is already quite big).
It is correct that this would reduce the number of connections, but I don't think they are that expensive. Do you have any information here?

File: connectionprivate.cpp (hunk 1 above)
Comment (KitsuneRal) [inline connectionprivate.cpp]: This is just a matter of personal taste, as for me. Since you drive the car, I willingly comply :)

File: room.cpp (hunk 2 above)
Comment (KitsuneRal) [inline room.cpp]: Ok, I knew there will be questions regarding it... Actually this hack is quite well-known - and even considered more transparent than if/else if. People mention that the thing is essentially a "glorified goto", no matter how you avoid using goto (a quick googling gave me that: https://stackoverflow.com/questions/243967/do-you-consider-this-technique-bad). Having said that, I can rewrite it in your preferred style.

**Comment by KitsuneRal** [comment]
Regarding signal connections - basically, it's memory that we are wasting. Since connections are stored in hash map, access time doesn't really change when this map grows. As for memory - each connection is at least four pointers big, which (with 32 bit pointers) brings at least 4x4x(number of connections) bytes needed to store the map (not counting the map's own overhead). Probably, 8Mb+ is not so much these days - but still looks quite a waste.
And yeah, Connection is a bit of a god class already. Gotta do something with it.
To summarise, I don't have a clear solution here. I'd still prefer to save on multiplying connections, but it needs some mediator class (if not Connection).

File: room.cpp (hunk 2 above)
Comment (KitsuneRal) [inline room.cpp]: On another thought - I think I'd just make one more function and replace break's with return's.

**Comment by KitsuneRal** [comment]
I rebased and amended the branch according to the minor comments. Regarding connections and Connection (pun intended :) ) - let's discuss separately (it might be a good idea to separate this part into another PR).

**Comment by KitsuneRal** [comment]
Reworked the branch once more to remove two commits about Connection:: signals and connections to them. A PR for organising signals for user events will come separately, and probably not soon - I'll think how to consolidate signals through another mediator than Connection. Therefore, the current PR should be clear to merge with respect to all your comments.

**Comment by KitsuneRal** [comment]
Here's one more commit to take care of user renamings. Turned out to be quite non-trivial; the good news though is that solving this problem also solved the question regarding signals mediator. As it turns out, every room has to process each renaming anyway because it causes potential disambiguations (even for users that did not rename themselves). So the best thing to do here for consolidation is to throw all user renaming signals into all rooms which is quite a bit of signal-slot spam (and it's worse than having additional megabytes of memory consumed instead). So, there you have it. The updated PR for the Quaternion part will come later.

**Comment by Fxrh** [comment]
Looks all good now :) Thanks for your work!
